File: Examples/xoShiRo128plus.js
export class XoShiRo128plus {
    constructor(seed) {
        this.s = new Uint32Array(4);

        let x = seed >>> 0; // force unsigned 32-bit

        for (let i = 0; i < 4; i++) {
            const { updatedX, output } = XoShiRo128plus._splitMix32(x);
            x = updatedX;       // carry forward updated state
            this.s[i] = output; // scrambled output
        }

        // Ensure state is not all zero
        if (this.s[0] === 0 && this.s[1] === 0 && this.s[2] === 0 && this.s[3] === 0) {
            this.s[0] = 1;
        }
    }

    static _splitMix32(x) {
        x = (x + 0x9E3779B9) >>> 0; // Step 1: increment state
        let z = x;
        z = Math.imul(z ^ (z >>> 16), 0x85EBCA6B) >>> 0; // Step 2
        z = Math.imul(z ^ (z >>> 13), 0xC2B2AE35) >>> 0; // Step 3
        return {
            updatedX: x,
            output: (z ^ (z >>> 16)) >>> 0 // Step 4: final scrambled value
        };
    }

    // Rotate left 32-bit
    static _rotl32(x, k) {
        return ((x << k) | (x >>> (32 - k))) >>> 0;
    }

    // Generate next uint
    nextUint() {
        const result = (this.s[0] + this.s[3]) >>> 0;

        const t = (this.s[1] << 9) >>> 0;

        this.s[2] ^= this.s[0];
        this.s[3] ^= this.s[1];
        this.s[1] ^= this.s[2];
        this.s[0] ^= this.s[3];

        this.s[2] ^= t;

        this.s[3] = XoShiRo128plus._rotl32(this.s[3], 11);

        return result >>> 0;
    }

    // Generate float in [0, 1)
    nextFloat() {
        return (this.nextUint() >>> 8) / (1 << 24);
    }

    // Generate boolean
    nextBool() {
        return (this.nextUint() & 1) === 1;
    }
}

File: MapConfiguratorCivilization7/Common/XoShiRo128plus.cs

namespace MapConfiguratorCivilization7.Helper
{
    public class XoShiRo128plus
    {
        private uint[] s = new uint[4];

        public XoShiRo128plus(int seed)
        {
            // Use SplitMix32 to fill s[0..3] from a single 32-bit seed
            uint x = (uint)seed;
            for (int i = 0; i < 4; i++)
            {
                s[i] = SplitMix32(ref x);
            }

            // Ensure state is not all zero
            if (s[0] == 0 && s[1] == 0 && s[2] == 0 && s[3] == 0)
            {
                s[0] = 1;
            }
        }

        private static uint SplitMix32(ref uint x)
        {
            uint z = (x += 0x9E3779B9);
            z = (z ^ (z >> 16)) * 0x85EBCA6B;
            z = (z ^ (z >> 13)) * 0xC2B2AE35;
            return z ^ (z >> 16);
        }

        private static uint Rotl(uint x, int k)
        {
            return (x << k) | (x >> (32 - k));
        }

        public uint NextUint()
        {
            uint result = s[0] + s[3];

            uint t = s[1] << 9;

            s[2] ^= s[0];
            s[3] ^= s[1];
            s[1] ^= s[2];
            s[0] ^= s[3];

            s[2] ^= t;

            s[3] = Rotl(s[3], 11);

            return result;
        }

        public float NextFloat()
        {
            return (NextUint() >> 8) / (float)(1 << 24); // 24-bit float precision [0,1)
        }

        public bool NextBool()
        {
            return (NextUint() & 1) == 1;
        }
    }
}
